File: monitor.js
const axios = require("axios");
const Database = require("better-sqlite3");
const https = require("https");
const fs = require("node:fs");
const path = require("node:path");

// Ignore SSL certificate errors
const agent = new https.Agent({
  rejectUnauthorized: false,
});

class GreekSiteMonitor {
  constructor(dbPath = "site_status.db") {
    this.dbPath = dbPath;
    this.timeout = 10000; // 10 seconds
    this.db = new Database(dbPath);
    this.db.pragma("journal_mode = WAL");

    // Initialize database
    this.initializeDb();

    // Axios instance with custom config
    this.axiosInstance = axios.create({
      timeout: this.timeout,
      httpsAgent: agent,
      maxRedirects: 3,
      validateStatus: function (status) {
        return status >= 200 && status < 600;
      },
    });
    // List of important Greek websites to monitor
    this.sites = {
      "gov.gr": "https://www.gov.gr",
      gsis: "https://www.gsis.gr",
      efka: "https://www.efka.gov.gr",
      ktimalogio: "https://www.ktimatologio.gr",
      "et.gr": "https://www.et.gr",
      oaed: "https://www.dypa.gov.gr",
      immigration: "https://migration.gov.gr",
      passport: "https://www.passport.gov.gr",
      eopyy: "https://www.eopyy.gov.gr",

      // Ministries
      ministry_digital: "https://mindigital.gr",
      ministry_finance: "https://www.minfin.gr",
      ministry_foreign: "https://www.mfa.gr",
      ministry_interior: "https://www.ypes.gr",
      ministry_education: "https://www.minedu.gov.gr",
      ministry_health: "https://www.moh.gov.gr",
      ministry_justice: "https://www.ministryofjustice.gr",
      ministry_culture: "https://www.culture.gov.gr",
      ministry_tourism: "https://mintour.gov.gr",

      // Education
      eudoxus: "https://eudoxus.gr",
      myschool: "https://myschool.sch.gr",
      uoa: "https://www.uoa.gr",
      auth: "https://www.auth.gr",
      ntua: "https://www.ntua.gr",
      upatras: "https://www.upatras.gr",

      // Transportation
      oasa: "https://www.oasa.gr",
      trainose: "https://www.hellenictrain.gr",
      athens_airport: "https://www.aia.gr",
      oasth: "https://oasth.gr",

      // Utilities & Services
      eydap: "https://www.eydap.gr",
      elta: "https://www.elta.gr",
      cosmote: "https://www.cosmote.gr",
      nova: "https://www.nova.gr",
      vodafone: "https://www.vodafone.gr",

      // Emergency Services
      ekav: "https://www.ekab.gr",
      civilprotection: "https://www.civilprotection.gr",
      fireservice: "https://www.fireservice.gr",
      astynomia: "https://www.astynomia.gr",

      // Banking
      bankofgreece: "https://www.bankofgreece.gr",
      nbg: "https://www.nbg.gr",
      alpha: "https://www.alpha.gr",
      eurobank: "https://www.eurobank.gr",

      // News & Media
      ert: "https://www.ert.gr",
      kathimerini: "https://www.kathimerini.gr",
      tovima: "https://www.tovima.gr",
      naftemporiki: "https://www.naftemporiki.gr",
      "in.gr": "https://www.in.gr",

      // Weather & Environment
      emy: "https://www.emy.gr",
      meteo: "https://www.meteo.gr",
      noa: "https://www.noa.gr",

      // Sports
      gga: "https://gga.gov.gr",
      epo: "https://www.epo.gr",
      oaka: "https://www.oaka.com.gr",
    };
  }

  initializeDb() {
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS site_status (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        site_name TEXT,
        url TEXT,
        status_code INTEGER,
        response_time REAL,
        is_up INTEGER,
        error_message TEXT,
        timestamp DATETIME
      )
    `);

    // Prepare statements for better performance
    this.insertStmt = this.db.prepare(`
      INSERT INTO site_status 
      (site_name, url, status_code, response_time, is_up, error_message, timestamp)
      VALUES (?, ?, ?, ?, ?, ?, ?)
    `);

    this.getLatestStmt = this.db.prepare(`
      SELECT * FROM site_status
      WHERE id IN (
        SELECT MAX(id)
        FROM site_status
        GROUP BY site_name
      )
      ORDER BY site_name
    `);
  }

  async checkSite(siteName, urlConfig) {
    const startTime = Date.now();
    let url, options;

    if (typeof urlConfig === "string") {
      url = urlConfig;
      options = {};
    } else {
      url = urlConfig.url;
      options = {
        maxRedirects: urlConfig.maxRedirects || 5,
        specialHandling: urlConfig.specialHandling || false,
      };
    }

    try {
      const response = await this.axiosInstance.get(url, {
        ...options,
        maxRedirects: options.maxRedirects,
      });

      const responseTime = (Date.now() - startTime) / 1000;

      return {
        site_name: siteName,
        url,
        status_code: response.status,
        response_time: responseTime,
        is_up: response.status >= 200 && response.status < 400 ? 1 : 0, // Convert boolean to integer
        error_message: null,
        timestamp: new Date().toISOString(),
      };
    } catch (error) {
      const errorMessage = this.formatError(error);
      console.error(`Error checking ${siteName} (${url}):`, errorMessage);

      return {
        site_name: siteName,
        url,
        status_code: error.response?.status || null,
        response_time: null,
        is_up: 0, // Convert boolean to integer
        error_message: errorMessage,
        timestamp: new Date().toISOString(),
      };
    }
  }

  formatError(error) {
    if (error.code === "ERR_FR_TOO_MANY_REDIRECTS") return "Too many redirects";
    if (error.code === "ECONNREFUSED") return "Connection refused";
    if (error.code === "ECONNABORTED") return "Connection timed out";
    if (error.code === "ENOTFOUND") return "DNS lookup failed";
    return error.message || "Unknown error";
  }

  saveStatus(status) {
    this.insertStmt.run(
      status.site_name,
      status.url,
      status.status_code,
      status.response_time,
      status.is_up,
      status.error_message,
      status.timestamp
    );
  }

  async checkAllSites() {
    const batchSize = 5;
    const sites = Object.entries(this.sites);

    // Use a transaction for better performance
    const transaction = this.db.transaction((statuses) => {
      for (const status of statuses) {
        this.saveStatus(status);
      }
    });

    for (let i = 0; i < sites.length; i += batchSize) {
      const batch = sites.slice(i, i + batchSize);
      const promises = batch.map(([siteName, urlConfig]) =>
        this.checkSite(siteName, urlConfig)
      );

      const results = await Promise.allSettled(promises);
      const validResults = results
        .filter((result) => result.status === "fulfilled")
        .map((result) => result.value);

      // Execute transaction with batch results
      transaction(validResults);

      // Log results
      for (const status of validResults) {
        console.log(
          `Checked ${status.site_name}: ${status.is_up ? "UP" : "DOWN"}${
            status.error_message ? ` (${status.error_message})` : ""
          }`
        );
      }
    }
  }

  getLatestStatus() {
    const results = this.getLatestStmt.all();
    // Convert integer back to boolean for display
    return results.map((result) => ({
      ...result,
      is_up: result.is_up === 1,
    }));
  }

  async startScheduledMonitoring(intervalMinutes = 1) {
    console.log(
      `Starting scheduled monitoring every ${intervalMinutes} minute(s)...`
    );

    // Initial run
    await this.runMonitoringCycle();

    // Schedule subsequent runs
    setInterval(async () => {
      await this.runMonitoringCycle();
    }, intervalMinutes * 60 * 1000);
  }

  async runMonitoringCycle() {
    const startTime = new Date();
    console.log(
      `\nStarting monitoring cycle at ${startTime.toLocaleString("en-US", {
        timeZone: "Europe/Athens",
      })}`
    );

    try {
      await this.checkAllSites();

      console.log("\nLatest status for all sites:");
      const latestStatus = this.getLatestStatus();

      // Format the status data
      const statusData = {
        lastUpdated: new Date().toISOString(),
        console: latestStatus,
        sites: latestStatus.map((status) => ({
          name: status.site_name,
          url: status.url,
          isUp: status.is_up,
          responseTime: status.response_time,
          statusCode: status.status_code,
          error: status.error_message || null,
          lastChecked: status.timestamp,
          // Add category based on the site name
          category: this.getSiteCategory(status.site_name),
        })),
      };

      // Save to JSON file
      const publicDir = path.join(__dirname, "public");
      if (!fs.existsSync(publicDir)) {
        fs.mkdirSync(publicDir);
      }
      fs.writeFileSync(
        path.join(publicDir, "latest-status.json"),
        JSON.stringify(statusData, null, 2)
      );

      console.table(
        latestStatus.map((status) => ({
          Site: status.site_name,
          Status: status.is_up ? "UP" : "DOWN",
          "Response Time": status.response_time
            ? `${status.response_time.toFixed(2)}s`
            : "N/A",
          Error: status.error_message || "None",
          "Last Checked": new Date(status.timestamp).toLocaleString("en-US", {
            timeZone: "Europe/Athens",
          }),
        }))
      );

      const endTime = new Date();
      const duration = (endTime - startTime) / 1000;
      console.log(
        `\nMonitoring cycle completed in ${duration.toFixed(2)} seconds`
      );
    } catch (error) {
      console.error("Error during monitoring cycle:", error);
    }
  }

  // Helper method to determine site category
  getSiteCategory(siteName) {
    const categories = {
      government: [
        "gov.gr",
        "gsis",
        "efka",
        "ktimalogio",
        "et.gr",
        "oaed",
        "immigration",
        "passport",
        "eopyy",
      ],
      ministries: [
        "ministry_digital",
        "ministry_finance",
        "ministry_foreign",
        "ministry_interior",
        "ministry_education",
        "ministry_health",
        "ministry_justice",
        "ministry_culture",
        "ministry_tourism",
      ],
      education: ["eudoxus", "myschool", "uoa", "auth", "ntua", "upatras"],
      transportation: ["oasa", "trainose", "athens_airport", "oasth"],
      utilities: ["eydap", "elta", "cosmote", "nova", "vodafone"],
      emergency: ["ekav", "civilprotection", "fireservice", "astynomia"],
      banking: ["bankofgreece", "nbg", "alpha", "eurobank"],
      media: ["ert", "kathimerini", "tovima", "naftemporiki", "in.gr"],
      weather: ["emy", "meteo", "noa"],
      sports: ["gga", "epo", "oaka"],
    };

    for (const [category, sites] of Object.entries(categories)) {
      if (sites.includes(siteName)) {
        return category;
      }
    }
    return "other";
  }
  shutdown() {
    try {
      if (this.db) {
        // Close the database connection
        this.db.close();
        console.log("Database connection closed successfully");
      }
    } catch (error) {
      console.error("Error while shutting down:", error);
    }
  }
}

// Example usage with scheduled monitoring
let monitor = null;

async function main() {
  monitor = new GreekSiteMonitor();

  try {
    // Start monitoring every 1 minute
    await monitor.startScheduledMonitoring(1);
  } catch (error) {
    console.error("Error in main:", error);
    if (monitor) monitor.shutdown();
    process.exit(1);
  }

  // Handle graceful shutdown
  process.on("SIGINT", () => {
    console.log("\nReceived SIGINT. Closing database and exiting...");
    if (monitor) monitor.shutdown();
    process.exit(0);
  });

  process.on("SIGTERM", () => {
    console.log("\nReceived SIGTERM. Closing database and exiting...");
    if (monitor) monitor.shutdown();
    process.exit(0);
  });
}

if (require.main === module) {
  main().catch(console.error);
}

module.exports = GreekSiteMonitor;

File: monitor.py
import urllib.request
import urllib.error
import http.client
import ssl
import json
import time
from datetime import datetime
import threading
from typing import Dict, List, Union, Optional
from dataclasses import dataclass
import logging
import socket

@dataclass
class SiteConfig:
    url: str
    max_redirects: int = 5
    special_handling: bool = False

@dataclass
class MonitoringResult:
    site_name: str
    url: str
    status_code: Optional[int]
    response_time: Optional[float]
    is_up: int
    error_message: Optional[str]
    timestamp: str

class GreekSiteMonitorClient:
    def __init__(self, server_url: str = "http://localhost:3000"):
        self.server_url = server_url
        self.timeout = 10  # 10 seconds
        
        # List of important Greek websites to monitor
        self.sites: Dict[str, Union[str, SiteConfig]] = {
            "gov.gr": "https://www.gov.gr",
            "gsis": "https://www.gsis.gr",
            "efka": "https://www.efka.gov.gr",
            # ... rest of the sites remain the same
        }
        
        # Configure logging
        logging.basicConfig(
            level=logging.INFO,
            format='%(asctime)s - %(levelname)s - %(message)s'
        )
        self.logger = logging.getLogger(__name__)

        # Configure SSL context
        self.ssl_context = ssl.create_default_context()
        self.ssl_context.check_hostname = False
        self.ssl_context.verify_mode = ssl.CERT_NONE

    def format_error(self, error: Exception) -> str:
        """Format common network errors into readable messages."""
        error_str = str(error)
        if isinstance(error, urllib.error.URLError):
            if isinstance(error.reason, socket.timeout):
                return "Connection timed out"
            return str(error.reason)
        elif isinstance(error, socket.gaierror):
            return "DNS lookup failed"
        elif isinstance(error, http.client.RemoteDisconnected):
            return "Connection closed by remote server"
        return error_str or "Unknown error"

    def check_site(self, site_name: str, url_config: Union[str, SiteConfig]) -> MonitoringResult:
        """Check a single site's status."""
        start_time = time.time()
        
        if isinstance(url_config, str):
            url = url_config
            max_redirects = 5
        else:
            url = url_config.url
            max_redirects = url_config.max_redirects

        opener = urllib.request.build_opener(
            urllib.request.HTTPRedirectHandler(),
            urllib.request.HTTPSHandler(context=self.ssl_context)
        )
        opener.addheaders = [('User-Agent', 'PythonMonitorClient/1.0')]

        try:
            urllib.request.install_opener(opener)
            response = opener.open(url, timeout=self.timeout)
            response_time = time.time() - start_time
            
            return MonitoringResult(
                site_name=site_name,
                url=url,
                status_code=response.status,
                response_time=response_time,
                is_up=1 if 200 <= response.status < 400 else 0,
                error_message=None,
                timestamp=datetime.now().isoformat()
            )

        except Exception as error:
            error_message = self.format_error(error)
            self.logger.error(f"Error checking {site_name} ({url}): {error_message}")
            
            return MonitoringResult(
                site_name=site_name,
                url=url,
                status_code=None,
                response_time=None,
                is_up=0,
                error_message=error_message,
                timestamp=datetime.now().isoformat()
            )

    def send_results_to_server(self, results: List[MonitoringResult]) -> None:
        """Send monitoring results to the server."""
        try:
            # Commented out as per original version
            # data = json.dumps([vars(result) for result in results]).encode('utf-8')
            # request = urllib.request.Request(
            #     f"{self.server_url}/api/status",
            #     data=data,
            #     headers={'Content-Type': 'application/json'}
            # )
            # with urllib.request.urlopen(request) as response:
            #     response.read()
            self.logger.info("Results sent to server successfully")
        except Exception as error:
            self.logger.error(f"Error sending results to server: {error}")
            raise

    def check_all_sites(self) -> List[MonitoringResult]:
        """Check all sites in batches."""
        batch_size = 5
        sites = list(self.sites.items())
        all_results = []

        for i in range(0, len(sites), batch_size):
            batch = sites[i:i + batch_size]
            batch_results = []
            
            # Use threads for parallel processing within batch
            threads = []
            for site_name, url_config in batch:
                thread = threading.Thread(
                    target=lambda: batch_results.append(
                        self.check_site(site_name, url_config)
                    )
                )
                thread.start()
                threads.append(thread)
            
            # Wait for all threads in batch to complete
            for thread in threads:
                thread.join()
            
            all_results.extend(batch_results)

            # Log results
            for status in batch_results:
                self.logger.info(
                    f"Checked {status.site_name}: "
                    f"{'UP' if status.is_up else 'DOWN'}"
                    f"{f' ({status.error_message})' if status.error_message else ''}"
                )

        return all_results

    def run_monitoring_cycle(self) -> None:
        """Run a complete monitoring cycle."""
        start_time = datetime.now()
        self.logger.info(
            f"\nStarting monitoring cycle at {start_time.strftime('%Y-%m-%d %H:%M:%S')}"
        )

        try:
            results = self.check_all_sites()
            self.send_results_to_server(results)

            duration = (datetime.now() - start_time).total_seconds()
            self.logger.info(f"\nMonitoring cycle completed in {duration:.2f} seconds")
        except Exception as error:
            self.logger.error(f"Error during monitoring cycle: {error}")

    def start_scheduled_monitoring(self, interval_minutes: float = 1) -> None:
        """Start scheduled monitoring with specified interval."""
        self.logger.info(f"Starting scheduled monitoring every {interval_minutes} minute(s)...")

        try:
            while True:
                self.run_monitoring_cycle()
                time.sleep(interval_minutes * 60)
        except KeyboardInterrupt:
            self.logger.info("\nMonitoring stopped by user")
        except Exception as error:
            self.logger.error(f"Monitoring stopped due to error: {error}")

if __name__ == "__main__":
    client = GreekSiteMonitorClient()
    client.start_scheduled_monitoring(1)